akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -1772,14 +1804,29 @@ trait OutHandler {
   @throws(classOf[Exception])
   def onPull(): Unit
 
+  // Hack to make sure that old `onDownstreamFinish` can be called without losing the cause in the default implementation
+  private[this] var _lastCancellationCause: Throwable = _
+
   /**
    * Called when the output port will no longer accept any new elements. After this callback no other callbacks will
    * be called for this port.
    */
   @throws(classOf[Exception])
-  def onDownstreamFinish(): Unit = {
-    GraphInterpreter.currentInterpreter.activeStage.completeStage()
-  }
+  // FIXME: we should add this deprecation but first we need to fix all our own usages
+  // @deprecatedOverride("Override method that provides cause.", since = "2.6.0")
+  def onDownstreamFinish(): Unit =
+    GraphInterpreter.currentInterpreter.activeStage.cancelStage(_lastCancellationCause)
```

Comment:
Perhaps add a check here that `_lastCancellationCause` is not `null`, to catch people calling `onDownstreamFinished` from user code?

Reply:
Good suggestion.

Reply:
Done